lib/ips/register.cpp

```diff
@@ -0,0 +1,106 @@
+
+bool Register::init() { return true; }
+
+bool Register::check() {
+
+  logger->debug("Checking register interface: Base address: 0x{:08x}",
+                getBaseAddr(registerMemory));
+  uint32_t buf;
+  // we shouldn't change the rate register, because this can lead to hardware fault, so start at 1
+  for (size_t i = 1; i < registerNum; i++) {
+    setRegister(i, static_cast<uint32_t>(i));
+  }
+
+  for (size_t i = 1; i < registerNum; i++) {
+    buf = getRegister(i);
+    if (buf != i) {
+      logger->error("Register {}: 0x{:08x} != 0x{:08x}", i, buf, i);
+      return false;
+    }
+  }
+
+  resetAllRegisters();
+
+  for (size_t i = 0; i < registerNum; i++) {
+    logger->trace("Register {}: 0x{:08x}", i, getRegister(i));
+  }
+
+  // This is Dino specific for now - we should possibly move this to Dino in the future
+  uint32_t rate = getRegister(0);
+  float scale = getRegisterFloat(1);
+  float offset = getRegisterFloat(2);
+  logger->info("Check: Register configuration: Rate: {}, Scale: {}, Offset: {}",
+               rate, scale, offset);
+
+  return true;
+}
+
+void Register::setRegister(size_t reg, uint32_t value) {
+  if (reg >= registerNum) {
+    logger->error("Register index out of range: {}/{}", reg, registerNum);
+    throw std::out_of_range("Register index out of range");
+  }
+  Xil_Out32(getBaseAddr(registerMemory) + REGISTER_OUT(reg), value);
+}
+
+void Register::setRegister(size_t reg, float value) {
+  if (reg >= registerNum) {
+    logger->error("Register index out of range: {}/{}", reg, registerNum);
+    throw std::out_of_range("Register index out of range");
+  }
+  Xil_Out32(getBaseAddr(registerMemory) + REGISTER_OUT(reg),
+            reinterpret_cast<uint32_t &>(value));
+}
+
+uint32_t Register::getRegister(size_t reg) {
+  if (reg >= registerNum) {
+    logger->error("Register index out of range: {}/{}", reg, registerNum);
+    throw std::out_of_range("Register index out of range");
+  }
+  return Xil_In32(getBaseAddr(registerMemory) + REGISTER_OUT(reg));
+}
+
+float Register::getRegisterFloat(size_t reg) {
+  if (reg >= registerNum) {
+    logger->error("Register index out of range: {}/{}", reg, registerNum);
+    throw std::out_of_range("Register index out of range");
+  }
+  uint32_t value = Xil_In32(getBaseAddr(registerMemory) + REGISTER_OUT(reg));
+  return reinterpret_cast<float &>(value);
+}
+
+void Register::resetRegister(size_t reg) {
+  if (reg >= registerNum) {
+    logger->error("Register index out of range: {}/{}", reg, registerNum);
+    throw std::out_of_range("Register index out of range");
+  }
+  Xil_Out32(getBaseAddr(registerMemory) + REGISTER_RESET, (1 << reg));
+}
+
+void Register::resetAllRegisters() {
+  Xil_Out32(getBaseAddr(registerMemory) + REGISTER_RESET, 0xFFFFFFFF);
+}
+
+Register::~Register() {}
+
+static char n[] = "register";
+static char d[] = "Register interface VHDL module 'registerif'";
+static char v[] = "xilinx.com:module_ref:registerif:";
```

Comment:
Is this an official IP block from Xilinx?

If not, I think it would be better to rename its identifier to something like `de.rwth-aachen.eonerc.acs:module_ref:registerif`.

I am also wondering what the purpose of the driver is? Arn't most IP drivers controlled via some sort of register interface? Is this intended to be a base class for all those IPs?

Reply:
This is pretty similar to the GPIO block but a bit more versatil. It could be integrated in other blocks but for now I need it separate. I don't think we can rename RTL Blocks.